clippy_lints/src/index_refutable_slice.rs

```diff
@@ -0,0 +1,284 @@
+
+            let found_slices = find_slice_values(cx, let_pat);
+            if !found_slices.is_empty();
+            let filtered_slices = filter_lintable_slices(cx, found_slices, self.indexing_limit, if_then);
+            if !filtered_slices.is_empty();
+            then {
+                for slice in filtered_slices.values() {
+                    lint_slice(cx, slice);
+                }
+            }
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn find_slice_values(cx: &LateContext<'_>, pat: &hir::Pat<'_>) -> FxHashMap<hir::HirId, SliceLintInformation> {
+    let mut removed_pat: FxHashSet<hir::HirId> = FxHashSet::default();
+    let mut slices: FxHashMap<hir::HirId, SliceLintInformation> = FxHashMap::default();
+    pat.walk_always(|pat| {
+        if let hir::PatKind::Binding(binding, value_hir_id, ident, sub_pat) = pat.kind {
+            // We'll just ignore mut and ref mut for simplicity sake right now
+            if let hir::BindingAnnotation::Mutable | hir::BindingAnnotation::RefMut = binding {
+                return;
+            }
+
+            // This block catches bindings with sub patterns. It would be hard to build a correct suggestion
+            // for them and it's likely that the user knows what they are doing in such a case.
+            if removed_pat.contains(&value_hir_id) {
+                return;
+            }
+            if sub_pat.is_some() {
+                removed_pat.insert(value_hir_id);
+                slices.remove(&value_hir_id);
+                return;
+            }
+
+            let bound_ty = cx.typeck_results().node_type(pat.hir_id);
+            if let ty::Slice(inner_ty) | ty::Array(inner_ty, _) = bound_ty.peel_refs().kind() {
+                // The values need to use the `ref` keyword if they can't be copied.
+                // This will need to be adjusted if the lint want to support multable access in the future
+                let src_is_ref = bound_ty.is_ref() && binding != hir::BindingAnnotation::Ref;
+                let is_copy = cx
+                    .tcx
+                    .lang_items()
+                    .copy_trait()
+                    .map_or(false, |trait_id| implements_trait(cx, inner_ty, trait_id, &[]));
+                let needs_ref = !(src_is_ref || is_copy);
+
+                let slice_info = slices
+                    .entry(value_hir_id)
+                    .or_insert_with(|| SliceLintInformation::new(ident, needs_ref));
+                slice_info.pattern_spans.push(pat.span);
+            }
+        }
+    });
+
+    slices
+}
+
+fn lint_slice(cx: &LateContext<'_>, slice: &SliceLintInformation) {
+    let used_indices = slice
+        .index_use
+        .iter()
+        .map(|(index, _)| *index)
+        .collect::<FxHashSet<_>>();
+
+    let value_name = |index| format!("{}_{}", slice.ident.name, index);
+
+    if let Some(max_index) = used_indices.iter().max() {
```

Comment:
You already checked `slice.index_use` is not empty so you can unwrap `max()`. Maybe move the empty check to this function.

Reply:
I found no simple way to move the `is_empty` check to this function. I usually try to avoid calling unwrap in Clippy code, and  here I think it's fine to add one level of indention. Is that alright with you? Otherwise, an `unwrap` call with an explanatory comment is also fine IMO

Reply:
I was thinking to just remove the `retain` call but maybe I'm missing something. Yeah this is fine as is.

Reply:
Alright, I missed that simple solution. Updated the code to remove the retain statement :upside_down_face:

Reply:
Well now `filter_lintable_slices` doesn't filter anything 😆. Maybe rename to `find_slice_usages`.

Totally optional, but you could pass it `&mut found_slices` and use `expr_visitor` instead of `impl Visitor`.

Reply:
The visitor still filters slices that are used for operations other than index operations. The remove method can be called in line 272.

Reply:
Ah okay!